File: go-service/internal/application/usecase/create_product.go
package usecase

import (
	"context"

	"github.com/google/uuid"

	appcache "github.com/example/interview-stack/go-service/internal/application/cache"
	"github.com/example/interview-stack/go-service/internal/application/dto"
	"github.com/example/interview-stack/go-service/internal/domain"
)

// CreateProductUseCase encapsulates business logic for creating products.
type CreateProductUseCase struct {
	repo         domain.ProductRepository
	logger       domain.Logger
	clock        domain.Clock
	queue        domain.WriteQueue
	writeBehind  bool
	source       string
	cacheWriter  cacheWriter
	cacheEnabled bool
}

func NewCreateProductUseCase(
	repo domain.ProductRepository,
	logger domain.Logger,
	clock domain.Clock,
	cacheSvc *appcache.Service,
	queue domain.WriteQueue,
	writeBehind bool,
	source string,
) *CreateProductUseCase {
	return &CreateProductUseCase{
		repo:         repo,
		logger:       logger,
		clock:        clock,
		queue:        queue,
		writeBehind:  writeBehind,
		source:       source,
		cacheWriter:  newCacheWriter(cacheSvc, logger),
		cacheEnabled: cacheSvc != nil && cacheSvc.Enabled(),
	}
}

func (uc *CreateProductUseCase) Execute(ctx context.Context, input dto.CreateProductRequest) (*dto.ProductResponse, error) {
	product := input.ToDomain()
	if err := product.Validate(); err != nil {
		return nil, err
	}
	now := uc.clock.Now()
	product.CreatedAt = now
	product.UpdatedAt = now

	if uc.writeBehind && uc.queue != nil {
		if product.ID == "" {
			product.ID = uuid.NewString()
		}
		dtoProduct := dto.FromDomainProduct(&product)
		uc.cacheWriter.upsertProduct(ctx, &dtoProduct)
		uc.cacheWriter.invalidateList(ctx)
		eventPayload := product
		event := domain.WriteEvent{
			Type:      domain.WriteEventCreate,
			ID:        product.ID,
			Payload:   &eventPayload,
			Timestamp: now,
			Source:    uc.source,
		}
		if err := uc.queue.Enqueue(ctx, event); err != nil {
			return nil, domain.NewInternalError("failed to enqueue create event", err)
		}
		uc.logger.Info("product enqueued for creation", domain.KV("id", product.ID))
		return &dtoProduct, nil
	}

	created, err := uc.repo.CreateProduct(ctx, product)
	if err != nil {
		return nil, err
	}
	dtoProduct := dto.FromDomainProduct(created)
	if uc.cacheEnabled {
		uc.cacheWriter.upsertProduct(ctx, &dtoProduct)
		uc.cacheWriter.invalidateList(ctx)
	}
	uc.logger.Info("product created", domain.KV("id", created.ID))
	return &dtoProduct, nil
}

File: node-service/internal/application/usecases/createProduct.js
const { v4: uuid } = require('uuid');
const { toDomain, toResponse } = require('../dto/productDto');

class CreateProductUseCase {
  /**
   * @param {import('../../domain/productRepository')} repository
   * @param {import('../../domain/logger')} logger
   * @param {import('../../core/clock/Clock')} clock
   * @param {import('../services/cacheCoordinator')} cacheCoordinator
   * @param {import('../services/writeQueue')} writeQueue
   */
  constructor(repository, logger, clock, cacheCoordinator, writeQueue, writeBehindEnabled, source) {
    this.repository = repository;
    this.logger = logger;
    this.clock = clock;
    this.cache = cacheCoordinator;
    this.queue = writeQueue;
    this.writeBehind = Boolean(writeBehindEnabled);
    this.source = source || 'node-service';
  }

  async execute(input) {
    const product = toDomain(input);
    product.validate();
    const now = this.clock.now();
    product.createdAt = now;
    product.updatedAt = now;

    if (this.writeBehind && this.queue) {
      product.id = product.id || uuid();
      const response = toResponse(product);
      await this.cache?.upsertProduct(response);
      await this.enqueueEvent('create', product, now);
      this.logger.info('product enqueued for creation', { id: product.id });
      return response;
    }

    const created = await this.repository.create(product);
    const response = toResponse(created);
    await this.cache?.upsertProduct(response);
    this.logger.info('product created', { id: created.id });
    return response;
  }

  async enqueueEvent(type, product, timestamp) {
    const event = {
      type,
      id: product.id,
      payload: product,
      timestamp: timestamp?.toISOString?.() || new Date().toISOString(),
      source: this.source
    };
    await this.queue.enqueue(event);
  }
}

module.exports = CreateProductUseCase;
